File: elixir/ltx/lib/interplanet_ltx/session.ex
# session.ex — LTX v1.1 session state machine (Epic 72).
# Mirrors typescript/ltx/src/session.ts (Stories 68.1/68.2 —
# LTX-SPECIFICATION.md §5 plan lock, DEGRADED, quorum; §5.4 delay violations).
#
# Pure and time-injected: every event carries "nowMs", transition/2 never
# reads a clock, and all side-effects are returned as effect maps.

defmodule InterplanetLtx.Session do
  @moduledoc """
  Pure session state machine over string-keyed plan maps.

  States: DRAFT LOCKING LOCKED ACTIVE DEGRADED EMERGENCY_HOLD COMPLETE ABORTED.
  Events are string-keyed maps with a "type" key (matching the conformance
  vectors), e.g. `%{"type" => "START_LOCK", "nowMs" => 1000}`.
  """

  alias InterplanetLtx.Constants
  alias InterplanetLtx.Segments

  # ── Helpers ─────────────────────────────────────────────────────────────────

  defp participants(plan) do
    Enum.filter(plan["nodes"] || [], fn n -> n["role"] == "PARTICIPANT" end)
  end

  @doc "2 × one-way delay to the furthest node, in ms (LTX-SPECIFICATION.md §5.1)."
  def lock_timeout_ms(plan) do
    max_delay_s =
      Enum.reduce(plan["nodes"] || [], 0, fn n, m -> max(m, n["delay"] || 0) end)

    Constants.default_plan_lock_timeout_factor() * max_delay_s * 1000
  end

  defp quorum_count(plan, quorum) do
    total = length(participants(plan))

    cond do
      quorum in [:majority, "majority"] -> div(total, 2) + 1
      is_integer(quorum) -> quorum |> max(1) |> min(total)
      true -> total
    end
  end

  # Ascending-delay fallback ordering over confirmed participants (§5.3).
  defp confirmed_subset(ctx) do
    host = List.first(ctx.plan["nodes"])

    confirmed =
      participants(ctx.plan)
      |> Enum.filter(fn n -> ctx.confirmations[n["id"]] == ctx.plan_id end)
      |> Enum.sort_by(fn n -> n["delay"] || 0 end)
      |> Enum.map(fn n -> n["id"] end)

    [host["id"] | confirmed]
  end

  defp full_lock_reached(ctx) do
    Enum.all?(participants(ctx.plan), fn n ->
      ctx.confirmations[n["id"]] == ctx.plan_id
    end)
  end

  defp quorum_reached(ctx) do
    confirmed =
      participants(ctx.plan)
      |> Enum.count(fn n -> ctx.confirmations[n["id"]] == ctx.plan_id end)

    confirmed >= ctx.quorum_threshold
  end

  # Declared one-way delay for a node: v3 pair matrix HOST row, else node delay.
  defp declared_delay_s(plan, node_id) do
    node = Enum.find(plan["nodes"] || [], fn n -> n["id"] == node_id end)

    cond do
      is_nil(node) ->
        nil

      is_map(plan["delays"]) ->
        host_id = List.first(plan["nodes"])["id"]
        key = [host_id, node_id] |> Enum.sort() |> Enum.join("|")

        if is_number(plan["delays"][key]),
          do: plan["delays"][key],
          else: node["delay"] || 0

      true ->
        node["delay"] || 0
    end
  end

  # ── Session construction ────────────────────────────────────────────────────

  @doc """
  Create a session context in DRAFT state. `plan_id` is supplied by the
  caller (`Segments.make_plan_id/1`) so this module stays pure.
  Options: `quorum: :all | :majority | integer` (default `:all`).
  """
  def create_session(plan, plan_id, opts \\ []) do
    plan = Segments.plan_map(plan)

    %{
      state: "DRAFT",
      plan: plan,
      plan_id: plan_id,
      session_root_plan_id: plan_id,
      plan_version: plan["planVersion"] || 1,
      lock: nil,
      lock_started_at_ms: nil,
      lock_timeout_ms: lock_timeout_ms(plan),
      confirmations: %{},
      mismatched: [],
      quorum_threshold: quorum_count(plan, Keyword.get(opts, :quorum, :all)),
      subset: nil,
      degraded_reasons: [],
      resume_state: nil,
      pending_amendment: nil
    }
  end

  # ── Transition plumbing ─────────────────────────────────────────────────────

  defp moved(ctx, to, event, effects, detail \\ nil) do
    entry =
      %{
        type: "state_transition",
        from: ctx.state,
        to: to,
        event: event["type"],
        at_ms: event["nowMs"]
      }
      |> then(fn e -> if detail, do: Map.put(e, :detail, detail), else: e end)

    %{ctx: %{ctx | state: to}, effects: [%{kind: "audit", entry: entry} | effects]}
  end

  defp unchanged(ctx, effects \\ []), do: %{ctx: ctx, effects: effects}

  defp invalid(ctx, event) do
    %{kind: "notify", level: "warn", code: "INVALID_EVENT",
      detail: "#{event["type"]} ignored in state #{ctx.state}"}
  end

  defp degrade(ctx, event, reason, extra \\ []) do
    next = %{ctx | degraded_reasons: ctx.degraded_reasons ++ [reason]}

    effects =
      [
        %{kind: "notify", level: "warn", code: "DEGRADED", detail: reason},
        %{kind: "escalate", code: "DEGRADED", detail: reason}
      ] ++ extra

    if ctx.state == "DEGRADED" do
      unchanged(next, Enum.take(effects, 1))
    else
      moved(next, "DEGRADED", event, effects, reason)
    end
  end

  # ── Transition function ─────────────────────────────────────────────────────

  @doc """
  Advance the session state machine. Pure: same (ctx, event) always yields
  the same result. Returns `%{ctx: ctx, effects: effects}`.
  """
  def transition(ctx, %{"type" => "START_LOCK"} = event) do
    if ctx.state != "DRAFT" do
      unchanged(ctx, [invalid(ctx, event)])
    else
      host_id = List.first(ctx.plan["nodes"])["id"]

      next = %{ctx |
        lock_started_at_ms: event["nowMs"],
        confirmations: Map.put(ctx.confirmations, host_id, ctx.plan_id)
      }

      moved(next, "LOCKING", event, [])
    end
  end

  def transition(ctx, %{"type" => "PLAN_CONFIRM"} = event) do
    if ctx.state not in ["LOCKING", "DEGRADED"] do
      unchanged(ctx, [invalid(ctx, event)])
    else
      node_id = event["nodeId"]
      plan_id = event["planId"]
      next = %{ctx | confirmations: Map.put(ctx.confirmations, node_id, plan_id)}

      cond do
        plan_id != ctx.plan_id ->
          next = %{next | mismatched: Enum.reject(ctx.mismatched, &(&1 == node_id)) ++ [node_id]}

          unchanged(next, [%{
            kind: "notify", level: "warn", code: "PLANID_MISMATCH",
            detail: "#{node_id} confirmed #{plan_id}, expected #{ctx.plan_id} (resolve per §5.5)"
          }])

        true ->
          next = %{next | mismatched: Enum.reject(ctx.mismatched, &(&1 == node_id))}

          if full_lock_reached(next) do
            locked = %{next | lock: "FULL", subset: nil}
            # Late full confirmation recovers a DEGRADED quorum lock (§5.2).
            moved(locked, "LOCKED", event, [%{
              kind: "notify", level: "info", code: "LOCKED", detail: "full lock achieved"
            }])
          else
            unchanged(next)
          end
      end
    end
  end

  def transition(ctx, %{"type" => "TICK"} = event) do
    cond do
      ctx.state != "LOCKING" -> unchanged(ctx)
      is_nil(ctx.lock_started_at_ms) -> unchanged(ctx)
      event["nowMs"] - ctx.lock_started_at_ms < ctx.lock_timeout_ms -> unchanged(ctx)
      # Lock timeout expired (§5.1).
      quorum_reached(ctx) ->
        subset = confirmed_subset(ctx)
        next = %{ctx | lock: "QUORUM", subset: subset}

        missing =
          participants(ctx.plan)
          |> Enum.filter(fn n -> ctx.confirmations[n["id"]] != ctx.plan_id end)
          |> Enum.map(fn n -> n["id"] end)

        degrade(next, event,
          "quorum lock with subset [#{Enum.join(subset, ",")}]; unconfirmed: [#{Enum.join(missing, ",")}]")

      true ->
        degrade(ctx, event, "plan-lock timeout without quorum")
    end
  end

  def transition(ctx, %{"type" => "SESSION_START"} = event) do
    cond do
      ctx.state == "LOCKED" ->
        moved(ctx, "ACTIVE", event, [])

      ctx.state == "DEGRADED" and not is_nil(ctx.lock) ->
        # §5.2: escalation to HOST required before TX.
        unchanged(ctx, [%{
          kind: "escalate", code: "DEGRADED_START",
          detail: "session start requested while DEGRADED; HOST decision required"
        }])

      true ->
        unchanged(ctx, [invalid(ctx, event)])
    end
  end

  def transition(ctx, %{"type" => "DELAY_MEASURED"} = event) do
    if ctx.state not in ["ACTIVE", "LOCKED", "DEGRADED"] do
      unchanged(ctx)
    else
      declared = declared_delay_s(ctx.plan, event["nodeId"])
      measured = event["measuredDelayS"]

      cond do
        is_nil(declared) ->
          unchanged(ctx, [invalid(ctx, event)])

        abs(measured - declared) > Constants.delay_violation_degraded_s() ->
          degrade(ctx, event,
            "delay violation #{event["nodeId"]}: measured #{measured}s vs declared #{declared}s (>#{Constants.delay_violation_degraded_s()}s)")

        abs(measured - declared) > Constants.delay_violation_warn_s() ->
          unchanged(ctx, [%{
            kind: "notify", level: "warn", code: "DELAY_VIOLATION",
            detail: "#{event["nodeId"]}: measured #{measured}s vs declared #{declared}s"
          }])

        true ->
          unchanged(ctx)
      end
    end
  end

  def transition(ctx, %{"type" => "EOK_OVERRIDE"} = event) do
    cond do
      ctx.state in ["COMPLETE", "ABORTED"] ->
        unchanged(ctx)

      event["verified"] != true ->
        unchanged(ctx, [%{
          kind: "notify", level: "error", code: "OVERRIDE_REJECTED",
          detail: event["reason"] || "override failed verification"
        }])

      ctx.state == "EMERGENCY_HOLD" ->
        unchanged(ctx)

      true ->
        next = %{ctx | resume_state: ctx.state}

        moved(next, "EMERGENCY_HOLD", event, [%{
          kind: "notify", level: "error", code: "EMERGENCY_HOLD",
          detail: event["reason"] || "verified EOK override"
        }])
    end
  end

  def transition(ctx, %{"type" => "AMENDMENT_PROPOSED"} = event) do
    cond do
      ctx.state not in ["ACTIVE", "LOCKED", "DEGRADED"] ->
        unchanged(ctx, [invalid(ctx, event)])

      event["planVersion"] != ctx.plan_version + 1 ->
        unchanged(ctx, [%{
          kind: "notify", level: "error", code: "AMENDMENT_REJECTED",
          detail: "planVersion #{event["planVersion"]} != #{ctx.plan_version} + 1"
        }])

      true ->
        # Delta re-lock (§6.4): timeout scoped to the furthest affected node.
        affected_ids = event["affectedNodeIds"] || []

        max_delay_s =
          (ctx.plan["nodes"] || [])
          |> Enum.filter(fn n -> n["id"] in affected_ids end)
          |> Enum.reduce(0, fn n, m -> max(m, n["delay"] || 0) end)

        pending = %{
          plan_id: event["planId"],
          plan_version: event["planVersion"],
          affected_node_ids: affected_ids,
          confirmed: [],
          proposed_at_ms: event["nowMs"],
          timeout_ms: Constants.default_plan_lock_timeout_factor() * max_delay_s * 1000
        }

        unchanged(%{ctx | pending_amendment: pending}, [%{
          kind: "notify", level: "info", code: "AMENDMENT_PROPOSED",
          detail: "plan #{event["planId"]} v#{event["planVersion"]}; awaiting [#{Enum.join(affected_ids, ",")}]"
        }])
    end
  end

  def transition(ctx, %{"type" => "AMENDMENT_CONFIRMED"} = event) do
    pa = ctx.pending_amendment

    cond do
      is_nil(pa) or event["planId"] != pa.plan_id ->
        unchanged(ctx, [invalid(ctx, event)])

      event["nodeId"] not in pa.affected_node_ids ->
        unchanged(ctx)

      true ->
        confirmed =
          Enum.reject(pa.confirmed, &(&1 == event["nodeId"])) ++ [event["nodeId"]]

        if length(confirmed) < length(pa.affected_node_ids) do
          unchanged(%{ctx | pending_amendment: %{pa | confirmed: confirmed}})
        else
          # All affected nodes confirmed — the amendment applies. The caller
          # swaps ctx.plan for the verified successor plan.
          next = %{ctx |
            plan_id: pa.plan_id,
            plan_version: pa.plan_version,
            pending_amendment: nil
          }

          unchanged(next, [%{
            kind: "notify", level: "info", code: "AMENDMENT_APPLIED",
            detail: "plan #{pa.plan_id} v#{pa.plan_version} in effect (root #{ctx.session_root_plan_id})"
          }])
        end
    end
  end

  def transition(ctx, %{"type" => "HOST_DECISION"} = event) do
    cond do
      event["decision"] == "abort" ->
        if ctx.state in ["COMPLETE", "ABORTED"] do
          unchanged(ctx)
        else
          moved(ctx, "ABORTED", event, [])
        end

      event["decision"] == "resume" and ctx.state == "EMERGENCY_HOLD" ->
        back = ctx.resume_state || "ACTIVE"
        moved(%{ctx | resume_state: nil}, back, event, [])

      event["decision"] == "continue" and ctx.state == "DEGRADED" ->
        # §5.2: HOST elects to continue with the confirmed subset.
        moved(ctx, "ACTIVE", event, [%{
          kind: "notify", level: "warn", code: "CONTINUE_DEGRADED",
          detail:
            if(ctx.subset,
              do: "continuing with subset [#{Enum.join(ctx.subset, ",")}]",
              else: "continuing despite degraded condition")
        }])

      true ->
        unchanged(ctx, [invalid(ctx, event)])
    end
  end

  def transition(ctx, %{"type" => "SESSION_END"} = event) do
    if ctx.state in ["ACTIVE", "DEGRADED"] do
      moved(ctx, "COMPLETE", event, [])
    else
      unchanged(ctx, [invalid(ctx, event)])
    end
  end
end

File: python/ltx/src/interplanet_ltx/_session.py
"""
interplanet_ltx._session — Session state machine
Story 68.5 — LTX-SPECIFICATION.md §5 (plan lock, DEGRADED, quorum) and §5.4
(delay-matrix violations). Mirrors typescript/ltx/src/session.ts.

Pure and time-injected: every event carries `now_ms`, transition() never reads
a clock, and all side-effects are returned as effect dicts. Plans are plain
dicts (the same shape sign_plan()/verify_plan() take).

States: DRAFT -> LOCKING -> LOCKED -> ACTIVE <-> DEGRADED -> COMPLETE,
plus EMERGENCY_HOLD (verified EOK override) and ABORTED.
"""

from typing import Any, Dict, List, Optional, Tuple

# Delay-matrix violation thresholds in seconds (LTX-SPECIFICATION.md §5.4)
DELAY_VIOLATION_WARN_S = 120
DELAY_VIOLATION_DEGRADE_S = 300

# Plan-lock timeout factor (LTX-SPECIFICATION.md §5.1)
LOCK_TIMEOUT_FACTOR = 2


def _participants(plan: Dict[str, Any]) -> List[Dict[str, Any]]:
    return [n for n in plan.get('nodes', []) if n.get('role') == 'PARTICIPANT']


def lock_timeout_ms(plan: Dict[str, Any]) -> int:
    """2 x one-way delay to the furthest node, in ms (§5.1)."""
    max_delay_s = max([n.get('delay', 0) for n in plan.get('nodes', [])] or [0])
    return int(LOCK_TIMEOUT_FACTOR * max_delay_s * 1000)


def _quorum_count(plan: Dict[str, Any], quorum: Any) -> int:
    total = len(_participants(plan))
    if quorum == 'majority':
        return total // 2 + 1
    if isinstance(quorum, int) and not isinstance(quorum, bool):
        return min(max(quorum, 1), total)
    return total  # 'all' (default)


def _confirmed_subset(ctx: Dict[str, Any]) -> List[str]:
    host = ctx['plan']['nodes'][0]
    confirmed = sorted(
        (n for n in _participants(ctx['plan'])
         if ctx['confirmations'].get(n['id']) == ctx['planId']),
        key=lambda n: n.get('delay', 0),
    )
    return [host['id']] + [n['id'] for n in confirmed]


def _full_lock_reached(ctx: Dict[str, Any]) -> bool:
    return all(ctx['confirmations'].get(n['id']) == ctx['planId']
               for n in _participants(ctx['plan']))


def _quorum_reached(ctx: Dict[str, Any]) -> bool:
    confirmed = sum(1 for n in _participants(ctx['plan'])
                    if ctx['confirmations'].get(n['id']) == ctx['planId'])
    return confirmed >= ctx['quorumThreshold']


def _declared_delay_s(plan: Dict[str, Any], node_id: str) -> Optional[float]:
    node = next((n for n in plan.get('nodes', []) if n.get('id') == node_id), None)
    if node is None:
        return None
    delays = plan.get('delays')
    if isinstance(delays, dict):
        host_id = plan['nodes'][0]['id']
        key = '|'.join(sorted([host_id, node_id]))
        if isinstance(delays.get(key), (int, float)):
            return delays[key]
    return node.get('delay', 0)


def create_session(plan: Dict[str, Any], plan_id: str,
                   quorum: Any = 'all') -> Dict[str, Any]:
    """Create a session context in DRAFT state. plan_id from make_plan_id()."""
    return {
        'state': 'DRAFT',
        'plan': plan,
        'planId': plan_id,
        'sessionRootPlanId': plan_id,
        'planVersion': plan.get('planVersion', 1),
        'lock': None,
        'lockStartedAtMs': None,
        'lockTimeoutMs': lock_timeout_ms(plan),
        'confirmations': {},
        'mismatched': [],
        'quorumThreshold': _quorum_count(plan, quorum),
        'subset': None,
        'degradedReasons': [],
        'resumeState': None,
        'pendingAmendment': None,
    }


def _moved(ctx: Dict[str, Any], to: str, event: Dict[str, Any],
           effects: List[Dict[str, Any]], detail: str = '') -> Tuple[Dict[str, Any], List[Dict[str, Any]]]:
    entry: Dict[str, Any] = {
        'type': 'state_transition',
        'from': ctx['state'],
        'to': to,
        'event': event['type'],
        'atMs': event['nowMs'],
    }
    if detail:
        entry['detail'] = detail
    next_ctx = dict(ctx)
    next_ctx['state'] = to
    return next_ctx, [{'kind': 'audit', 'entry': entry}] + effects


def _invalid(ctx: Dict[str, Any], event: Dict[str, Any]) -> Dict[str, Any]:
    return {'kind': 'notify', 'level': 'warn', 'code': 'INVALID_EVENT',
            'detail': f"{event['type']} ignored in state {ctx['state']}"}


def _degrade(ctx: Dict[str, Any], event: Dict[str, Any], reason: str,
             extra: Optional[List[Dict[str, Any]]] = None) -> Tuple[Dict[str, Any], List[Dict[str, Any]]]:
    next_ctx = dict(ctx)
    next_ctx['degradedReasons'] = ctx['degradedReasons'] + [reason]
    effects = [
        {'kind': 'notify', 'level': 'warn', 'code': 'DEGRADED', 'detail': reason},
        {'kind': 'escalate', 'code': 'DEGRADED', 'detail': reason},
    ] + (extra or [])
    if ctx['state'] == 'DEGRADED':
        return next_ctx, effects[:1]
    return _moved(next_ctx, 'DEGRADED', event, effects, reason)


def transition(ctx: Dict[str, Any], event: Dict[str, Any]) -> Tuple[Dict[str, Any], List[Dict[str, Any]]]:
    """Advance the session state machine (pure). Returns (next_ctx, effects)."""
    etype = event['type']

    if etype == 'START_LOCK':
        if ctx['state'] != 'DRAFT':
            return ctx, [_invalid(ctx, event)]
        host_id = ctx['plan']['nodes'][0]['id']
        next_ctx = dict(ctx)
        next_ctx['lockStartedAtMs'] = event['nowMs']
        next_ctx['confirmations'] = dict(ctx['confirmations'], **{host_id: ctx['planId']})
        return _moved(next_ctx, 'LOCKING', event, [])

    if etype == 'PLAN_CONFIRM':
        if ctx['state'] not in ('LOCKING', 'DEGRADED'):
            return ctx, [_invalid(ctx, event)]
        next_ctx = dict(ctx)
        next_ctx['confirmations'] = dict(ctx['confirmations'], **{event['nodeId']: event['planId']})
        if event['planId'] != ctx['planId']:
            next_ctx['mismatched'] = [i for i in ctx['mismatched'] if i != event['nodeId']] + [event['nodeId']]
            return next_ctx, [{
                'kind': 'notify', 'level': 'warn', 'code': 'PLANID_MISMATCH',
                'detail': f"{event['nodeId']} confirmed {event['planId']}, expected {ctx['planId']} (resolve per §5.5)",
            }]
        next_ctx['mismatched'] = [i for i in ctx['mismatched'] if i != event['nodeId']]
        if _full_lock_reached(next_ctx):
            next_ctx['lock'] = 'FULL'
            next_ctx['subset'] = None
            return _moved(next_ctx, 'LOCKED', event, [
                {'kind': 'notify', 'level': 'info', 'code': 'LOCKED', 'detail': 'full lock achieved'},
            ])
        return next_ctx, []

    if etype == 'TICK':
        if ctx['state'] != 'LOCKING' or ctx['lockStartedAtMs'] is None:
            return ctx, []
        if event['nowMs'] - ctx['lockStartedAtMs'] < ctx['lockTimeoutMs']:
            return ctx, []
        if _quorum_reached(ctx):
            subset = _confirmed_subset(ctx)
            next_ctx = dict(ctx)
            next_ctx['lock'] = 'QUORUM'
            next_ctx['subset'] = subset
            missing = [n['id'] for n in _participants(ctx['plan'])
                       if ctx['confirmations'].get(n['id']) != ctx['planId']]
            return _degrade(next_ctx, event,
                            f"quorum lock with subset [{','.join(subset)}]; unconfirmed: [{','.join(missing)}]")
        return _degrade(ctx, event, 'plan-lock timeout without quorum')

    if etype == 'SESSION_START':
        if ctx['state'] == 'LOCKED':
            return _moved(ctx, 'ACTIVE', event, [])
        if ctx['state'] == 'DEGRADED' and ctx['lock'] is not None:
            return ctx, [{
                'kind': 'escalate', 'code': 'DEGRADED_START',
                'detail': 'session start requested while DEGRADED; HOST decision required',
            }]
        return ctx, [_invalid(ctx, event)]

    if etype == 'DELAY_MEASURED':
        if ctx['state'] not in ('ACTIVE', 'LOCKED', 'DEGRADED'):
            return ctx, []
        declared = _declared_delay_s(ctx['plan'], event['nodeId'])
        if declared is None:
            return ctx, [_invalid(ctx, event)]
        deviation = abs(event['measuredDelayS'] - declared)
        if deviation > DELAY_VIOLATION_DEGRADE_S:
            return _degrade(ctx, event,
                            f"delay violation {event['nodeId']}: measured {event['measuredDelayS']}s "
                            f"vs declared {declared}s (>{DELAY_VIOLATION_DEGRADE_S}s)")
        if deviation > DELAY_VIOLATION_WARN_S:
            return ctx, [{
                'kind': 'notify', 'level': 'warn', 'code': 'DELAY_VIOLATION',
                'detail': f"{event['nodeId']}: measured {event['measuredDelayS']}s vs declared {declared}s",
            }]
        return ctx, []

    if etype == 'EOK_OVERRIDE':
        if ctx['state'] in ('COMPLETE', 'ABORTED'):
            return ctx, []
        if not event.get('verified'):
            return ctx, [{
                'kind': 'notify', 'level': 'error', 'code': 'OVERRIDE_REJECTED',
                'detail': event.get('reason', 'override failed verification'),
            }]
        if ctx['state'] == 'EMERGENCY_HOLD':
            return ctx, []
        next_ctx = dict(ctx)
        next_ctx['resumeState'] = ctx['state']
        return _moved(next_ctx, 'EMERGENCY_HOLD', event, [
            {'kind': 'notify', 'level': 'error', 'code': 'EMERGENCY_HOLD',
             'detail': event.get('reason', 'verified EOK override')},
        ])

    if etype == 'AMENDMENT_PROPOSED':
        if ctx['state'] not in ('ACTIVE', 'LOCKED', 'DEGRADED'):
            return ctx, [_invalid(ctx, event)]
        if event['planVersion'] != ctx['planVersion'] + 1:
            return ctx, [{
                'kind': 'notify', 'level': 'error', 'code': 'AMENDMENT_REJECTED',
                'detail': f"planVersion {event['planVersion']} != {ctx['planVersion']} + 1",
            }]
        affected = [n for n in ctx['plan']['nodes'] if n['id'] in event['affectedNodeIds']]
        max_delay_s = max([n.get('delay', 0) for n in affected] or [0])
        next_ctx = dict(ctx)
        next_ctx['pendingAmendment'] = {
            'planId': event['planId'],
            'planVersion': event['planVersion'],
            'affectedNodeIds': list(event['affectedNodeIds']),
            'confirmed': [],
            'proposedAtMs': event['nowMs'],
            'timeoutMs': int(LOCK_TIMEOUT_FACTOR * max_delay_s * 1000),
        }
        return next_ctx, [{
            'kind': 'notify', 'level': 'info', 'code': 'AMENDMENT_PROPOSED',
            'detail': f"plan {event['planId']} v{event['planVersion']}; "
                      f"awaiting [{','.join(event['affectedNodeIds'])}]",
        }]

    if etype == 'AMENDMENT_CONFIRMED':
        pa = ctx['pendingAmendment']
        if not pa or event['planId'] != pa['planId']:
            return ctx, [_invalid(ctx, event)]
        if event['nodeId'] not in pa['affectedNodeIds']:
            return ctx, []
        confirmed = [i for i in pa['confirmed'] if i != event['nodeId']] + [event['nodeId']]
        if len(confirmed) < len(pa['affectedNodeIds']):
            next_ctx = dict(ctx)
            next_ctx['pendingAmendment'] = dict(pa, confirmed=confirmed)
            return next_ctx, []
        next_ctx = dict(ctx)
        next_ctx['planId'] = pa['planId']
        next_ctx['planVersion'] = pa['planVersion']
        next_ctx['pendingAmendment'] = None
        return next_ctx, [{
            'kind': 'notify', 'level': 'info', 'code': 'AMENDMENT_APPLIED',
            'detail': f"plan {pa['planId']} v{pa['planVersion']} in effect "
                      f"(root {ctx['sessionRootPlanId']})",
        }]

    if etype == 'HOST_DECISION':
        decision = event['decision']
        if decision == 'abort':
            if ctx['state'] in ('COMPLETE', 'ABORTED'):
                return ctx, []
            return _moved(ctx, 'ABORTED', event, [])
        if decision == 'resume' and ctx['state'] == 'EMERGENCY_HOLD':
            next_ctx = dict(ctx)
            back = ctx['resumeState'] or 'ACTIVE'
            next_ctx['resumeState'] = None
            return _moved(next_ctx, back, event, [])
        if decision == 'continue' and ctx['state'] == 'DEGRADED':
            detail = (f"continuing with subset [{','.join(ctx['subset'])}]"
                      if ctx['subset'] else 'continuing despite degraded condition')
            return _moved(ctx, 'ACTIVE', event, [
                {'kind': 'notify', 'level': 'warn', 'code': 'CONTINUE_DEGRADED', 'detail': detail},
            ])
        return ctx, [_invalid(ctx, event)]

    if etype == 'SESSION_END':
        if ctx['state'] in ('ACTIVE', 'DEGRADED'):
            return _moved(ctx, 'COMPLETE', event, [])
        return ctx, [_invalid(ctx, event)]

    return ctx, [_invalid(ctx, event)]
